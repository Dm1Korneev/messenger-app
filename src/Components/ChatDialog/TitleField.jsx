import React from 'react';
import { TextField } from 'formik-material-ui';

const TitleField = (props) => (
  <TextField
    margin="normal"
    label="Title *"
    fullWidth
    name="title"
    {...props}
  />
);

export default TitleField;
