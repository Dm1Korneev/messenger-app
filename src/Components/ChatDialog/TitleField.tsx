import { TextField } from 'formik-material-ui';

type Props = Omit<
  React.ComponentPropsWithoutRef<typeof TextField>,
  'margin' | 'label' | 'fullWidth'>

const TitleField = (props: Props) => (
  <TextField
    margin="normal"
    label="Title *"
    fullWidth
    {...props}
  />
);

export default TitleField;
