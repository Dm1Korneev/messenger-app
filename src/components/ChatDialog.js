import React from "react";
import { ValidatorForm, TextValidator } from "react-material-ui-form-validator";
import PropTypes from "prop-types";

// components
import UsersAvatar from "./UsersAvatar";
import AvatarSelector from "./AvatarSelector";

// @material-ui
import withStyles from "@material-ui/core/styles/withStyles";
import Button from "@material-ui/core/Button";
import Dialog from "@material-ui/core/Dialog";
import DialogActions from "@material-ui/core/DialogActions";
import DialogContent from "@material-ui/core/DialogContent";
import DialogTitle from "@material-ui/core/DialogTitle";
import ListItem from "@material-ui/core/ListItem";
import List from "@material-ui/core/List";
import ListItemText from "@material-ui/core/ListItemText";

class ChatDialog extends React.Component {
  constructor(props) {
    super(props);
    const { isModify } = props;

    const state = {
      title: "",
      searchText: "",
      searchResult: [],
      selectedUserIds: [],
      titleIsModified: false,
      avatarIsModified: false,
      selectedUserIdsIsModified: false
    };

    if (isModify) {
      const { title, users } = props.chat;
      state.title = title;
      state.selectedUserIds = users;
    }
    this.state = state;
    this.avatarFileInput = React.createRef();
  }

  componentDidMount() {
    const { getUsers } = this.props;
    getUsers();
  }

  handleInputChange = event => {
    const target = event.target;
    const value = target.type === "checkbox" ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value,
      [name + "IsModified"]: true
    });
  };

  handleSearchTextChange = event => {
    const { users } = this.props;
    const searchText = event.target.value;

    if (!searchText) {
      this.setState({
        searchText,
        searchResult: []
      });
      return;
    } else {
      const searchResult = users.filter(
        value =>
          value.name.toUpperCase().includes(searchText.toUpperCase()) ||
          value.email.toUpperCase().includes(searchText.toUpperCase())
      );
      this.setState({
        searchText,
        searchResult
      });
    }
  };

  avatarOnChange = () => {
    this.setState({
      avatarIsModified: true
    });
  };

  handleSubmit = event => {
    const {
      onAddChat,
      onSaveChat,
      closeChatDialog,
      isModify,
      chat
    } = this.props;
    const {
      title,
      selectedUserIds,
      titleIsModified,
      avatarIsModified,
      selectedUserIdsIsModified
    } = this.state;
    const avatar = this.avatarFileInput.current.files[0];

    if (isModify) {
      let options = {};
      if (titleIsModified) {
        options = { ...options, title };
      }
      if (selectedUserIdsIsModified) {
        options = { ...options, users: selectedUserIds };
      }
      if (avatarIsModified) {
        options = {
          ...options,
          avatar: this.avatarFileInput.current.files[0]
        };
      }

      onSaveChat(chat._id, options);
    } else {
      onAddChat(title, avatar, selectedUserIds);
    }
    closeChatDialog();
  };

  userSelect = userId => {
    const selectedUserIdsFromState = this.state.selectedUserIds;

    let selectedUserIds;
    if (selectedUserIdsFromState.includes(userId)) {
      selectedUserIds = selectedUserIdsFromState.filter(
        element => element !== userId
      );
    } else {
      selectedUserIds = [...selectedUserIdsFromState, userId];
    }

    this.setState({
      selectedUserIds,
      selectedUserIdsIsModified: true
    });
  };

  render() {
    const { closeChatDialog, users, isModify, chat } = this.props;
    const { title, selectedUserIds, searchText, searchResult } = this.state;
    const usersList = !searchText ? users : searchResult;

    let avatar;
    if (isModify) {
      avatar = chat.avatar;
    }

    return (
      <Dialog open onClose={closeChatDialog}>
        <DialogTitle>{isModify ? "Modify chat" : "Add chat"}</DialogTitle>
        <DialogContent>
          <ValidatorForm id="validatorForm" onSubmit={this.handleSubmit}>
            <AvatarSelector
              onChange={this.avatarOnChange}
              avatar={avatar}
              avatarFileInput={this.avatarFileInput}
            />
            <TextValidator
              margin="normal"
              label="Title *"
              fullWidth
              onChange={this.handleInputChange}
              name="title"
              id="title"
              color="primary"
              value={title}
              validators={["required"]}
              errorMessages={["this field is required"]}
            />
            <TextValidator
              margin="normal"
              label="Search"
              fullWidth
              onChange={this.handleSearchTextChange}
              name="searchText"
              id="searchText"
              color="primary"
              value={searchText}
            />

            <List>
              {usersList.map(value => (
                <ListItem
                  button
                  key={value._id}
                  selected={selectedUserIds.includes(value._id)}
                  onClick={() => this.userSelect(value._id)}
                >
                  <UsersAvatar author={value.name} avatar={value.avatar} />
                  <ListItemText primary={`${value.name} (${value.email})`} />
                </ListItem>
              ))}
            </List>
          </ValidatorForm>
        </DialogContent>
        <DialogActions>
          <Button type="submit" color="primary" form="validatorForm">
            {isModify ? "Save" : "Add"}
          </Button>
          <Button onClick={closeChatDialog} color="primary">
            Close
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

ChatDialog.propTypes = {
  classes: PropTypes.object.isRequired,
  closeChatDialog: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
      email: PropTypes.string
    })
  ),
  isModify: PropTypes.bool.isRequired,
  chat: PropTypes.shape({
    avatar: PropTypes.string
  })
};

const styles = theme => ({});

export default withStyles(styles)(ChatDialog);
