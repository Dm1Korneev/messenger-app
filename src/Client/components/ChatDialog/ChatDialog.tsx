import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import LinearProgress from '@mui/material/LinearProgress';
import List from '@mui/material/List';
import ListItemAvatar from '@mui/material/ListItemAvatar';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemText from '@mui/material/ListItemText';
import TextField from '@mui/material/TextField';
import { Field, Form, Formik } from 'formik';
import { useEffect, useRef, useState } from 'react';
import * as Yup from 'yup';

import {
  name as nameValidation,
} from 'Client/common/validation';
import {
  useChatById, useUsers, useUpdateChat, UseUpdateChatPayload, useCreateChat,
} from 'Client/hooks';
import { User } from 'Client/types';

import { AvatarSelector } from '../AvatarSelector';
import { UsersAvatar } from '../UsersAvatar';

import TitleField from './TitleField';

const validationSchema = Yup.object().shape({
  title: nameValidation,
});

type FromValues = {
  title: string;
}

type ChatDialogProps = {
  chatId?: string
  onClose: ()=>void
  isModify?: boolean
}

export const ChatDialog = ({ chatId, onClose, isModify }: ChatDialogProps) => {
  const { data: users } = useUsers();
  const { data: chat } = useChatById(chatId);
  const { mutate: updateChat } = useUpdateChat();
  const { mutate: createChat } = useCreateChat();

  const [searchText, setSearchText] = useState<string>('');
  const [selectedUserIds, setSelectedUserIds] = useState<string[]>([]);
  const [searchResult, setSearchResult] = useState<User[]>([]);
  const [avatarIsModified, setAvatarIsModified] = useState<boolean>(false);

  const avatarFileInput = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (isModify && chat) {
      setSelectedUserIds(chat.users);
    }
  }, [chat, isModify]);

  const handleSearchTextChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSearchText = event.target.value;

    if (!newSearchText) {
      setSearchResult([]);
    } else {
      const newSearchResult = users.filter(
        (value) => value.name.toUpperCase().includes(newSearchText.toUpperCase())
          || value.email.toUpperCase().includes(newSearchText.toUpperCase()),
      );
      setSearchResult(newSearchResult);
    }
    setSearchText(newSearchText);
  };

  const avatarOnChange = () => {
    setAvatarIsModified(true);
  };

  const onSubmit = ({ title }: FromValues) => {
    const files = avatarFileInput?.current?.files;
    const avatar = files ? files[0] : undefined;

    if (isModify && chat) {
      const modifyData: UseUpdateChatPayload = {
        title,
        users: selectedUserIds,
      };
      if (avatarIsModified) {
        modifyData.avatar = avatar;
      }

      updateChat({ chatId: chat._id, modifyData });
    } else if (!isModify) {
      createChat({ title, avatar, users: selectedUserIds });
    }
    onClose();
  };

  const userSelect = (userId: string) => {
    const selectedUserIdsFromState = selectedUserIds;

    let newSelectedUserIds;
    if (selectedUserIdsFromState.includes(userId)) {
      newSelectedUserIds = selectedUserIdsFromState.filter(
        (element) => element !== userId,
      );
    } else {
      newSelectedUserIds = [...selectedUserIdsFromState, userId];
    }

    setSelectedUserIds(newSelectedUserIds);
  };

  const initialValues: FromValues = {
    title: (isModify ? chat?.title : '') ?? '',
  };

  const avatar = isModify ? chat?.avatar : undefined;

  const usersList = !searchText ? users : searchResult;
  return !isModify || chat ? (
    <Dialog onClose={onClose} open>
      <Formik
        initialValues={initialValues}
        onSubmit={onSubmit}
        validationSchema={validationSchema}
      >
        {({ isSubmitting, submitForm }) => (
          <Form>
            <DialogTitle>{isModify ? 'Modify chat' : 'Add chat'}</DialogTitle>
            <DialogContent>
              <AvatarSelector
                avatar={avatar}
                avatarFileInput={avatarFileInput}
                onChange={avatarOnChange}
              />
              <Field
                component={TitleField}
                name="title"
              />
              <TextField
                color="primary"
                fullWidth
                id="searchText"
                label="Search"
                margin="normal"
                name="searchText"
                onChange={handleSearchTextChange}
                value={searchText}
              />

              <List>
                {usersList.map((value) => (
                  <ListItemButton
                    key={value._id}
                    onClick={() => userSelect(value._id)}
                    selected={selectedUserIds.includes(value._id)}
                  >
                    <ListItemAvatar>
                      <UsersAvatar author={value.name} avatar={value.avatar} />
                    </ListItemAvatar>
                    <ListItemText primary={`${value.name} (${value.email})`} />
                  </ListItemButton>
                ))}
              </List>
              {isSubmitting && <LinearProgress />}
            </DialogContent>
            <DialogActions>
              <Button
                color="primary"
                disabled={isSubmitting}
                onClick={submitForm}
                type="submit"
              >
                {isModify ? 'Save' : 'Add'}
              </Button>
              <Button color="primary" onClick={onClose}>
                Close
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  ) : null;
};
