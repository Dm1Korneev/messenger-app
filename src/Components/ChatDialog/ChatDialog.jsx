import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import TextField from '@material-ui/core/TextField';
import LinearProgress from '@material-ui/core/LinearProgress';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';

import AvatarSelector from 'Components/AvatarSelector';
import UsersAvatar from 'Components/UsersAvatar';
import {
  name as nameValidation,
} from 'Common/validation';

import TitleField from './TitleField';

const validationSchema = Yup.object().shape({
  title: nameValidation,
});

const ChatDialog = ({
  users, onAddChat, onSaveChat, isModify, chat, getUsers, closeChatDialog,
}) => {
  const [searchText, setSearchText] = useState('');
  const [selectedUserIds, setSelectedUserIds] = useState([]);
  const [searchResult, setSearchResult] = useState([]);
  const [avatarIsModified, setAvatarIsModified] = useState(false);

  const avatarFileInput = useRef();

  useEffect(() => {
    getUsers();
  }, [getUsers]);

  useEffect(() => {
    if (isModify) {
      setSelectedUserIds(chat.users);
    }
  }, [chat, isModify]);

  const handleSearchTextChange = (event) => {
    const newSearchText = event.target.value;

    if (!newSearchText) {
      setSearchResult([]);
    } else {
      const newSearchResult = users.filter(
        (value) => value.name.toUpperCase().includes(newSearchText.toUpperCase())
          || value.email.toUpperCase().includes(newSearchText.toUpperCase()),
      );
      setSearchResult(newSearchResult);
    }
    setSearchText(newSearchText);
  };

  const avatarOnChange = () => {
    setAvatarIsModified(true);
  };

  const onSubmit = ({ title }) => {
    const avatar = avatarFileInput.current.files[0];

    if (isModify) {
      let options = { title, users: selectedUserIds };
      if (avatarIsModified) {
        options = {
          ...options,
          avatar: avatarFileInput.current.files[0],
        };
      }

      onSaveChat({ chatId: chat._id, options });
    } else {
      onAddChat({ title, avatar, selectedUserIds });
    }
    closeChatDialog();
  };

  const userSelect = (userId) => {
    const selectedUserIdsFromState = selectedUserIds;

    let newSelectedUserIds;
    if (selectedUserIdsFromState.includes(userId)) {
      newSelectedUserIds = selectedUserIdsFromState.filter(
        (element) => element !== userId,
      );
    } else {
      newSelectedUserIds = [...selectedUserIdsFromState, userId];
    }

    setSelectedUserIds(newSelectedUserIds);
  };

  let avatar;
  if (isModify) {
    avatar = chat.avatar;
  }

  const usersList = !searchText ? users : searchResult;
  return (
    <Dialog open onClose={closeChatDialog}>
      <Formik
        initialValues={{
          title: isModify ? chat.title : '',
        }}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
      >
        {({ isSubmitting, submitForm }) => (
          <Form>
            <DialogTitle>{isModify ? 'Modify chat' : 'Add chat'}</DialogTitle>
            <DialogContent>
              <AvatarSelector
                onChange={avatarOnChange}
                avatar={avatar}
                avatarFileInput={avatarFileInput}
              />
              <Field
                component={TitleField}
                name="title"
              />
              <TextField
                margin="normal"
                label="Search"
                fullWidth
                onChange={handleSearchTextChange}
                name="searchText"
                id="searchText"
                color="primary"
                value={searchText}
              />

              <List>
                {usersList.map((value) => (
                  <ListItem
                    button
                    key={value._id}
                    selected={selectedUserIds.includes(value._id)}
                    onClick={() => userSelect(value._id)}
                  >
                    <ListItemAvatar>
                      <UsersAvatar author={value.name} avatar={value.avatar} />
                    </ListItemAvatar>
                    <ListItemText primary={`${value.name} (${value.email})`} />
                  </ListItem>
                ))}
              </List>
              {isSubmitting && <LinearProgress />}
            </DialogContent>
            <DialogActions>
              <Button
                type="submit"
                color="primary"
                disabled={isSubmitting}
                onClick={submitForm}
              >
                {isModify ? 'Save' : 'Add'}
              </Button>
              <Button onClick={closeChatDialog} color="primary">
                Close
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  );
};

ChatDialog.defaultProps = {
  chat: undefined,
};

ChatDialog.propTypes = {
  closeChatDialog: PropTypes.func.isRequired,
  onAddChat: PropTypes.func.isRequired,
  onSaveChat: PropTypes.func.isRequired,
  getUsers: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(
    PropTypes.shape({
      _id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
      email: PropTypes.string,
    }),
  ).isRequired,
  isModify: PropTypes.bool.isRequired,
  chat: PropTypes.shape({
    _id: PropTypes.string,
    avatar: PropTypes.string,
    title: PropTypes.string,
    users: PropTypes.array,
  }),
};

export default ChatDialog;
