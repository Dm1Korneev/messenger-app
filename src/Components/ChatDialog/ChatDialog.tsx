import React, { useEffect, useRef, useState } from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import ListItem from '@material-ui/core/ListItem';
import List from '@material-ui/core/List';
import ListItemText from '@material-ui/core/ListItemText';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import TextField from '@material-ui/core/TextField';
import LinearProgress from '@material-ui/core/LinearProgress';
import { Field, Form, Formik } from 'formik';
import * as Yup from 'yup';
import { useDispatch, useSelector } from 'react-redux';

import { User } from 'Types';
import * as Actions from 'Redux/actions';
import { modifyChatDialogIsOpenSelector } from 'Selectors/session';
import { notCurrentUsersSelector } from 'Selectors/users';
import { modifiableChatSelector } from 'Selectors/chats';
import AvatarSelector from 'Components/AvatarSelector';
import UsersAvatar from 'Components/UsersAvatar';
import {
  name as nameValidation,
} from 'Common/validation';

import TitleField from './TitleField';

const validationSchema = Yup.object().shape({
  title: nameValidation,
});

type FromValues = {
  title: string;
}

const ChatDialog = () => {
  const dispatch = useDispatch();

  const users = useSelector(notCurrentUsersSelector);
  const isModify = useSelector(modifyChatDialogIsOpenSelector);
  const chat = useSelector(modifiableChatSelector);

  const [searchText, setSearchText] = useState<string>('');
  const [selectedUserIds, setSelectedUserIds] = useState<string[]>([]);
  const [searchResult, setSearchResult] = useState<User[]>([]);
  const [avatarIsModified, setAvatarIsModified] = useState<boolean>(false);

  const avatarFileInput = useRef<HTMLInputElement>(null);

  useEffect(() => {
    dispatch(Actions.getUsers());
  }, [dispatch]);

  useEffect(() => {
    if (isModify && chat) {
      setSelectedUserIds(chat.users);
    }
  }, [chat, isModify]);

  const closeChatDialog = () => dispatch(Actions.closeChatDialog());

  const handleSearchTextChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newSearchText = event.target.value;

    if (!newSearchText) {
      setSearchResult([]);
    } else {
      const newSearchResult = users.filter(
        (value) => value.name.toUpperCase().includes(newSearchText.toUpperCase())
          || value.email.toUpperCase().includes(newSearchText.toUpperCase()),
      );
      setSearchResult(newSearchResult);
    }
    setSearchText(newSearchText);
  };

  const avatarOnChange = () => {
    setAvatarIsModified(true);
  };

  const onSubmit = ({ title }: FromValues) => {
    const files = avatarFileInput?.current?.files;
    const avatar = files ? files[0] : undefined;

    if (isModify && chat) {
      const options: Actions.ModifyChatPayload['options'] = {
        title,
        users: selectedUserIds,
      };
      if (avatarIsModified) {
        options.avatar = avatar;
      }

      dispatch(Actions.modifyChat({ chatId: chat._id, options }));
    } else if (!isModify) {
      dispatch(Actions.createChat({ title, avatar, selectedUserIds }));
    }
    closeChatDialog();
  };

  const userSelect = (userId: string) => {
    const selectedUserIdsFromState = selectedUserIds;

    let newSelectedUserIds;
    if (selectedUserIdsFromState.includes(userId)) {
      newSelectedUserIds = selectedUserIdsFromState.filter(
        (element) => element !== userId,
      );
    } else {
      newSelectedUserIds = [...selectedUserIdsFromState, userId];
    }

    setSelectedUserIds(newSelectedUserIds);
  };

  const initialValues: FromValues = {
    title: (isModify ? chat?.title : '') ?? '',
  };

  const avatar = isModify ? chat?.avatar : undefined;

  const usersList = !searchText ? users : searchResult;
  return (
    <Dialog open onClose={closeChatDialog}>
      <Formik
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
      >
        {({ isSubmitting, submitForm }) => (
          <Form>
            <DialogTitle>{isModify ? 'Modify chat' : 'Add chat'}</DialogTitle>
            <DialogContent>
              <AvatarSelector
                onChange={avatarOnChange}
                avatar={avatar}
                avatarFileInput={avatarFileInput}
              />
              <Field
                component={TitleField}
                name="title"
              />
              <TextField
                margin="normal"
                label="Search"
                fullWidth
                onChange={handleSearchTextChange}
                name="searchText"
                id="searchText"
                color="primary"
                value={searchText}
              />

              <List>
                {usersList.map((value) => (
                  <ListItem
                    button
                    key={value._id}
                    selected={selectedUserIds.includes(value._id)}
                    onClick={() => userSelect(value._id)}
                  >
                    <ListItemAvatar>
                      <UsersAvatar author={value.name} avatar={value.avatar} />
                    </ListItemAvatar>
                    <ListItemText primary={`${value.name} (${value.email})`} />
                  </ListItem>
                ))}
              </List>
              {isSubmitting && <LinearProgress />}
            </DialogContent>
            <DialogActions>
              <Button
                type="submit"
                color="primary"
                disabled={isSubmitting}
                onClick={submitForm}
              >
                {isModify ? 'Save' : 'Add'}
              </Button>
              <Button onClick={closeChatDialog} color="primary">
                Close
              </Button>
            </DialogActions>
          </Form>
        )}
      </Formik>
    </Dialog>
  );
};

ChatDialog.defaultProps = {
  chat: undefined,
};

export default ChatDialog;
