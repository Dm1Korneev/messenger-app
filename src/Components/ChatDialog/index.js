import commonHoc from 'Components/commonHoc';
import { modifyChatDialogIsOpenSelector } from 'Selectors/session';
import { notCurrentUsersSelector } from 'Selectors/users';
import { modifiableChatSelector } from 'Selectors/chats';
import {
  closeChatDialog,
  createChat,
  getUsers,
  modifyChat,
} from 'Redux/actions';

import ChatDialog from './ChatDialog';

const mapStateToProps = (state) => ({
  users: notCurrentUsersSelector(state),
  isModify: modifyChatDialogIsOpenSelector(state),
  chat: modifiableChatSelector(state),
});

const mapDispatchToProps = {
  closeChatDialog,
  onAddChat: createChat,
  onSaveChat: modifyChat,
  getUsers,
};

export default commonHoc(ChatDialog, {
  mapStateToProps,
  mapDispatchToProps,
});

