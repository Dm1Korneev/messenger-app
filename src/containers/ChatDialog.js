import { connect } from "react-redux";
import {
  closeChatDialog,
  createChat,
  modifyChat,
  getUsers
} from "../redux/actions";
import ChatDialog from "../components/ChatDialog";

const mapStateToProps = state => {
  const usersFromState = state.users;
  const users = usersFromState.allIds
    .filter(element => element !== state.session.user._id)
    .map(value => usersFromState.byId[value]);
  const isModify = state.session.modifyChatDialogIsOpen;
  const chat = isModify && state.chats.byId[state.session.modifiableChat];

  return {
    users,
    isModify,
    chat
  };
};

const mapDispatchToProps = dispatch => ({
  closeChatDialog: () => dispatch(closeChatDialog()),
  onAddChat: (...args) => dispatch(createChat(...args)),
  onSaveChat: (...args) => dispatch(modifyChat(...args)),
  getUsers: () => dispatch(getUsers())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ChatDialog);
