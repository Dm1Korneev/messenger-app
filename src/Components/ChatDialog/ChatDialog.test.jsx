import React from 'react';

import ChatDialog from './ChatDialog';

jest.mock('Components/UsersAvatar', () => global.mockComponent('UsersAvatar'));
jest.mock('Components/AvatarSelector', () => global.mockComponent('AvatarSelector'));

jest.mock('@material-ui/core/Button', () => global.mockComponent('Button'));
jest.mock('@material-ui/core/Dialog', () => global.mockComponent('Dialog'));
jest.mock('@material-ui/core/DialogActions', () => global.mockComponent('DialogActions'));
jest.mock('@material-ui/core/DialogContent', () => global.mockComponent('DialogContent'));
jest.mock('@material-ui/core/DialogTitle', () => global.mockComponent('DialogTitle'));
jest.mock('@material-ui/core/ListItem', () => global.mockComponent('ListItem'));
jest.mock('@material-ui/core/List', () => global.mockComponent('List'));
jest.mock('@material-ui/core/ListItemText', () => global.mockComponent('ListItemText'));

const props = {
  closeChatDialog: jest.fn(),
  onAddChat: jest.fn(),
  onSaveChat: jest.fn(),
  getUsers: jest.fn(),
  users: [{
    _id: '_id',
    name: 'name',
    avatar: 'avatar',
    email: 'email',
  }],
  isModify: false,
  chat: {
    _id: '_id',
    avatar: 'avatar',
    title: 'title',
    users: [],
  },

};

describe('render ChatDialog component', () => {
  let wrapper;

  beforeAll(() => {
    wrapper = global.mount(
      <ChatDialog {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('ChatDialog').length).toBe(1);
  });

  test('Dialog subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'Dialog' }).length).toBe(1);
  });

  test('DialogContent subcomponent is render', () => {
    expect(wrapper.find({ originalcomponent: 'DialogContent' }).length).toBe(1);
  });
});

