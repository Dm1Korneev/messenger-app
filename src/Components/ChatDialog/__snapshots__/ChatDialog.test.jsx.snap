// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ChatDialog component snapshot test 1`] = `
<DocumentFragment>
  <div
    data-originalcomponent="Dialog"
    test-prop-open="true"
  >
    <div
      data-originalcomponent="DialogTitle"
    >
      Add chat
    </div>
    <div
      data-originalcomponent="DialogContent"
    >
      <div
        data-originalcomponent="ValidatorForm"
        test-prop-id="\\"validatorForm\\""
      >
        <div
          data-originalcomponent="AvatarSelector"
          test-prop-avatarfileinput="{\\"current\\":null}"
        />
        <div
          data-originalcomponent="TextValidator"
          test-prop-color="\\"primary\\""
          test-prop-errormessages="[\\"this field is required\\"]"
          test-prop-fullwidth="true"
          test-prop-id="\\"title\\""
          test-prop-label="\\"Title *\\""
          test-prop-margin="\\"normal\\""
          test-prop-name="\\"title\\""
          test-prop-validators="[\\"required\\"]"
          test-prop-value="\\"\\""
        />
        <div
          data-originalcomponent="TextValidator"
          test-prop-color="\\"primary\\""
          test-prop-fullwidth="true"
          test-prop-id="\\"searchText\\""
          test-prop-label="\\"Search\\""
          test-prop-margin="\\"normal\\""
          test-prop-name="\\"searchText\\""
          test-prop-value="\\"\\""
        />
        <div
          data-originalcomponent="List"
        >
          <div
            data-originalcomponent="ListItem"
            test-prop-button="true"
            test-prop-selected="false"
          >
            <div
              class="MuiListItemAvatar-root"
            >
              <div
                data-originalcomponent="UsersAvatar"
                test-prop-author="\\"name\\""
                test-prop-avatar="\\"avatar\\""
              />
            </div>
            <div
              data-originalcomponent="ListItemText"
              test-prop-primary="\\"name (email)\\""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      data-originalcomponent="DialogActions"
    >
      <div
        data-originalcomponent="Button"
        test-prop-color="\\"primary\\""
        test-prop-form="\\"validatorForm\\""
        test-prop-type="\\"submit\\""
      >
        Add
      </div>
      <div
        data-originalcomponent="Button"
        test-prop-color="\\"primary\\""
      >
        Close
      </div>
    </div>
  </div>
</DocumentFragment>
`;
